### PR TITLE
Salesforce Timeline Events

### PR DESCRIPTION
**Pull request by mqueme**
[//]: # ( Please answer the following questions: )

| Q  | A
| --- | ---
| Bug fix? | 
| New feature? | y
| Related user documentation PR URL |
| Related developer documentation PR URL | 
| Issues addressed (#s or URLs) |  https://github.com/mautic/mautic/issues/4854
| BC breaks? | 
| Deprecations? | 

[//]: # ( Note that all new features should have a related user and/or developer documentation PR in their respective repositories. )

[//]: # ( Required: )
#### Description:
The activity object used in Salesforce would only write email sends, form submits and point actions from the timeline. With this PR we are writing all the activities on the contact's timeline and we can filter through them.

[//]: # ( As applicable: )

#### Steps to test this PR:
1. follow the steps to set up the activity object from this link https://mautic.org/docs/en/plugins/salesforce.html **_Setting up Mautic's custom object in Salesforce_**
2. By default in order to keep BC, it'll send point changes, form submissions, and email read stats over to SF and Sugar. Once Salesforce's config form is updated with the events to include, it'll send those. 
3. Ensure that a Mautic contact is synced with a Salesforce Lead/Contact (run the m:i:s command)
4. Create a form with form submit actions to change the points and send an email to the contact
5. Submit the form in a guest/incognito browser with the synced contact's email
6. Open the email in a client outside of the browser you're logged into Mautic with
7. Run the `php app/console mautic:integration:pushactivity --integration Salesforce ` and all three activities should make it to the SF timeline
8. Edit the configuration form and select a few events to include, save, and then execute those events with a contact that's being synced. Then run the command. It should be synced with the SF timeline.

### CODE REVIEW

File: app/bundles/PluginBundle/Command/PushLeadActivityCommand.php [hunk 1]
```diff
@@ -54,6 +54,22 @@ protected function configure()
                 InputOption::VALUE_OPTIONAL,
                 'Send time interval to check updates on Salesforce, it should be a correct php formatted time interval in the past eg:(-10 minutes)'
             )
+            ->addOption(
+                '--include-events',
```
Comment (alanhartless) [inline app/bundles/PluginBundle/Command/PushLeadActivityCommand.php]: This needs to be a setting in the UI; not via the command itself

**Comment by alanhartless** [comment]
I ended up making a number of changes to this for several reason as mentioned in the commit messages above. Mainly 

1) UI/UX
2) BC breaks
3) Properly filter event data by the dates requested, send all applicable events (not just a limit of 100) and prevent memory overload from all the entities in Doctrine's memory from the engagements

**Comment by virlatinus** [comment]
Got the GROUP BY error again while executing `app/console m:i:s`:
```
[Doctrine\DBAL\Exception\DriverException]                                                                                        
  An exception occurred while executing 'SELECT DISTINCT(i.integration_entity_id), i.integration_entity, i.id, i.internal_entity_  
  id,l.email, l.lastname, l.company, l.firstname FROM integration_entity i INNER JOIN leads l ON l.id = i.internal_entity_id WHER  
  E (i.integration = ?) AND (i.integration_entity = ?) AND (i.internal_entity = ?) AND ((i.integration_entity_id IS NOT NULL) AND  
   (((i.last_sync_date IS NOT NULL) AND (l.date_modified > i.last_sync_date)) OR ((i.last_sync_date IS NULL) AND (l.date_modified  
   IS NOT NULL) AND (l.date_modified > l.date_added)))) AND (l.email IS NOT NULL) AND (l.date_modified BETWEEN ? and ?) GROUP BY   
  i.integration_entity, i.integration_entity_id LIMIT 100' with params ["Salesforce", "Contact", "lead", "2017-10-11 15:09:11", "  
  2017-10-11 15:24:11"]:                                                                                                           
  SQLSTATE[42000]: Syntax error or access violation: 1055 Expression #3 of SELECT list is not in GROUP BY clause and contains non  
  aggregated column 'mautic.i.id' which is not functionally dependent on columns in GROUP BY clause; this is incompatible with sq  
  l_mode=only_full_group_by
```
I have Mysql 5.7 and sql_mode `ONLY_FULL_GROUP_BY,STRICT_TRANS_TABLES,NO_ZERO_IN_DATE,NO_ZERO_DATE,ERROR_FOR_DIVISION_BY_ZERO,NO_AUTO_CREATE_USER,NO_ENGINE_SUBSTITUTION`

**Comment by mqueme** [comment]
This was fixed on a PR that has been merged @virlatinus

**Pull request merged**